ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+ * */
+public data class Interception<T : Any>(
+    val phase: PipelinePhase,
+    val action: (Pipeline<T, ApplicationCall>) -> Unit
+)
+
+/**
+ * Compatibility class. Interception class for Call phase
+ * */
+public typealias CallInterception = Interception<Unit>
+
+/**
+ * Compatibility class. Interception class for Receive phase
+ * */
+
+public typealias ReceiveInterception = Interception<ApplicationReceiveRequest>
+
+/**
+ * Compatibility class. Interception class for Send phase
+ * */
+public typealias ResponseInterception = Interception<Any>
+
+/**
+ * Compatibility class. Every plugin that needs to be compatible with `beforePlugin(...)` and `afterPlugin(...)`
+ * needs to implement this class. It defines a list of interceptions (see [Interception]) for phases in different pipelines
+ * that are being intercepted by the current plugin.
+ *
+ * It is needed in order to get first/last phase in each pipeline for the current plugin and create a new phase
+ * that is strictly earlier/later in this pipeline than any interception of the current plugin.
+ *
+ * Note: any [KtorPlugin] instance automatically fills these fields and there is no need to implement them by hand.
+ * But if you want to use old plugin API based on pipelines and phases and to make it available for `beforePlugin`/
+ * `afterPlugin` methods of the new plugins API, please consider filling the phases your plugin defines by implementing
+ * this interface [InterceptionsHolder].
+ *
+ * Please, use [defineInterceptions] method inside the first init block to simply define all the interceptions for your
+ * feature. Also it is useful to always delegate to [DefaultInterceptionsHolder] for simplicity.
+ * */
+public interface InterceptionsHolder {
+    /**
+     * A name for the current plugin.
+     * */
+    public val name: String get() = this.javaClass.simpleName
+
+    public val key: AttributeKey<out InterceptionsHolder>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val fallbackInterceptions: MutableList<CallInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val callInterceptions: MutableList<CallInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val monitoringInterceptions: MutableList<CallInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val beforeReceiveInterceptions: MutableList<ReceiveInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val onReceiveInterceptions: MutableList<ReceiveInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val beforeResponseInterceptions: MutableList<ResponseInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
+    public val onResponseInterceptions: MutableList<ResponseInterception>
+
+    @Deprecated("Please, use defineInterceptions instead")
```

Comment:
Why introduce a new API that is deprecated? If you need it to be public, but hidden from user, please use `@InternalApi`